rclrs/src/logging.rs

```diff
@@ -0,0 +1,342 @@
+// Copyright (c) 2019 Sequence Planner
+// SPDX-License-Identifier: Apache-2.0 AND MIT
+// Adapted from https://github.com/sequenceplanner/r2r/blob/89cec03d07a1496a225751159cbc7bfb529d9dd1/r2r/src/utils.rs
+// Further adapted from https://github.com/mvukov/rules_ros2/pull/371
+
+use std::{ffi::CString, sync::Mutex, time::Duration};
+
+use crate::rcl_bindings::*;
+
+// Used to protect calls to rcl/rcutils in case those functions manipulate global variables
+static LOG_GUARD: Mutex<()> = Mutex::new(());
+
+/// Calls the underlying rclutils logging function
+/// Don't call this directly, use the logging macros instead.
+///
+/// # Panics
+///
+/// This function might panic in the following scenarios:
+/// - A logger_name is provided that is not a valid c-string, e.g. contains extraneous null characters
+/// - The user-supplied "msg" is not a valid c-string, e.g. contains extraneous null characters
+/// - When called if the lock is already held by the current thread.
+/// - If the construction of CString objects used to create the log output fail,
+///   although, this highly unlikely to occur in most cases
+#[doc(hidden)]
+pub fn log(msg: &str, logger_name: &str, file: &str, line: u32, severity: LogSeverity) {
+    // currently not possible to get function name in rust.
+    // see https://github.com/rust-lang/rfcs/pull/2818
+    let function = CString::new("").unwrap();
+    let file = CString::new(file).unwrap();
+    let location = rcutils_log_location_t {
+        function_name: function.as_ptr(),
+        file_name: file.as_ptr(),
+        line_number: line as usize,
+    };
+    let format = CString::new("%s").unwrap();
+    let logger_name = CString::new(logger_name)
+        .expect("Logger name is a valid c style string, e.g. check for extraneous null characters");
+    let message = CString::new(msg)
+        .expect("Valid c style string required, e.g. check for extraneous null characters");
+    let severity = severity.to_native();
+
+    let _guard = LOG_GUARD.lock().unwrap();
```

Comment:
Since this is rclrs should use [`ENTITY_LIFECYCLE_MUTEX`](https://github.com/ros2-rust/ros2_rust/blob/f45a66f47dc727e3ccb13037a6c57923af1446c7/rclrs/src/context.rs#L23), which is the mutex that rclrs uses to ensure that global variables within the rcl layer and lower are protected.

rcl logging is notoriously not thread safe, so it's important to make sure we're locking appropriately.

Reply:
Good comment.  The thought had crossed my mind on this one - I thought it more akin to the mutex in rclpp logging, but I'm happy to change it to the `ENTITY_LIFECYCLE_MUTEX`

Reply:
Taken care of by https://github.com/geoff-imdex/ros2_rust/pull/6